pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/InvertedIndexAndDictionaryBasedForwardIndexCreator.java

```diff
@@ -0,0 +1,619 @@
+    LOGGER.info("Creating a new forward index for segment: {}, column: {}, isTemporary: {}", segmentName, _columnName,
+        _isTemporaryForwardIndex);
+
+    Map<String, String> metadataProperties;
+    if (_singleValue) {
+      metadataProperties = createForwardIndexForSVColumn();
+    } else {
+      metadataProperties = createForwardIndexForMVColumn();
+    }
+
+    LoaderUtils.writeIndexToV3Format(_segmentWriter, _columnName, _forwardIndexFile, ColumnIndexType.FORWARD_INDEX);
+
+    if (!_isTemporaryForwardIndex) {
+      // Only update the metadata and cleanup other indexes if the forward index to be created is permanent. If the
+      // forward index is temporary, it is meant to be used only for construction of other indexes and will be deleted
+      // once all the IndexHandlers have completed.
+      try {
+        LOGGER.info("Created forward index from inverted index and dictionary. Updating metadata properties for "
+            + "segment: {}, column: {}, property list: {}", segmentName, _columnName, metadataProperties);
+        ForwardIndexHandler.updateMetadataProperties(_segmentMetadata.getIndexDir(), metadataProperties);
+      } catch (Exception e) {
+        throw new IOException(
+            String.format("Failed to update metadata properties for segment: %s, column: %s", segmentName, _columnName),
+            e);
+      }
+
+      if (!_dictionaryEnabled) {
+        LOGGER.info("Clean up indexes no longer needed or which need to be rewritten for segment: {}, column: {}",
+            segmentName, _columnName);
+        // Delete the dictionary
+        _segmentWriter.removeIndex(_columnName, ColumnIndexType.DICTIONARY);
+
+        // We remove indexes that have to be rewritten when a dictEnabled is toggled. Note that the respective index
+        // handler will take care of recreating the index.
+        ForwardIndexHandler.removeDictRelatedIndexes(_columnName, _segmentWriter);
+      }
+    }
+
+    // Delete the marker file.
+    FileUtils.deleteQuietly(inProgress);
+
+    LOGGER.info("Created a new forward index for segment: {}, column: {}, isTemporary: {}", segmentName, _columnName,
+        _isTemporaryForwardIndex);
+  }
+
+  private Map<String, String> createForwardIndexForSVColumn()
+      throws IOException {
+    try (BitmapInvertedIndexReader invertedIndexReader =
+        (BitmapInvertedIndexReader) LoaderUtils.getInvertedIndexReader(_segmentWriter, _columnMetadata);
+        Dictionary dictionary = LoaderUtils.getDictionary(_segmentWriter, _columnMetadata)) {
+      boolean shouldTrackVarLengthMetadata = !_columnMetadata.getFieldSpec().getDataType().isFixedWidth();
+      int[] lengthOfLongestEntry = shouldTrackVarLengthMetadata ? new int[]{0} : new int[]{-1};
+      // Construct the forward index in the values buffer
+      for (int dictId = 0; dictId < _cardinality; dictId++) {
+        ImmutableRoaringBitmap docIdsBitmap = invertedIndexReader.getDocIds(dictId);
+        int finalDictId = dictId;
```

Comment:
Curious why do we need this temp assignment in `finalDictId` ? May be I am missing something